src/internal/operators/distinctUntilChanged.ts

```diff
@@ -1,78 +1,181 @@
+ * const changedHands$ = accountUpdates$.pipe(
+ *   distinctUntilChanged(undefined, update => update.updatedBy)
+ * );
  *
- * @param {function} [compare] Optional comparison function called to test if an item is distinct from the previous item in the source.
- * A return value of true indicates that it is the same, and a return value of false means they are different.
- * @return {Observable} An Observable that emits items from the source Observable with distinct values.
+ * changedHands$.subscribe(console.log);
+ * // Logs:
+ * // {updatedBy: "blesh", data: Array[0]}
+ * // {updatedBy: "ncjamieson", data: Array[0]}
+ * // {updatedBy: "blesh", data: Array[0]}
+ * ```
+ *
+ * @param comparator A function used to compare the previous and current keys for
+ * equality. Defaults to a `===` check.
+ * @param keySelector Used to select a key value to be passed to the `comparator`.
  */
-export function distinctUntilChanged<T, K>(compare?: (a: K, b: K) => boolean, keySelector?: (x: T) => K): MonoTypeOperatorFunction<T> {
-  compare = compare ?? defaultCompare;
+export function distinctUntilChanged<T, K>(
+  comparator: (previous: K, current: K) => boolean,
+  keySelector: (value: T) => K
+): MonoTypeOperatorFunction<T>;
+
+export function distinctUntilChanged<T, K>(
+  comparator?: (previous: K, current: K) => boolean,
+  keySelector: (value: T) => K = identity as (value: T) => K
+): MonoTypeOperatorFunction<T> {
+  // We've been allowing `null` do be passed as the `compare`, so we can't do
+  // a default value for the parameter, because that will only work
+  // for `undefined`.
+  comparator = comparator ?? defaultCompare;
+
   return operate((source, subscriber) => {
-    let prev: any;
+    // The previous key, used to compare against keys selected
+    // from new arrivals to determine "distinctiveness".
+    let previousKey: K;
+    // Whether or not this is the first value we've gotten.
     let first = true;
+
     source.subscribe(
       new OperatorSubscriber(subscriber, (value) => {
-        const key: any = keySelector ? keySelector(value) : value;
-        if (first || !compare!(prev, key)) {
+        // We always call the key selector.
+        const currentKey = keySelector(value);
+
+        // If it's the first value, we always emit it.
+        // Otherwise, we compare this key to the previous key, and
+        // if the comparer returns false, we emit.
+        if (first || !comparator!(previousKey, currentKey)) {
+          // Update our state *before* we emit the value
+          // as emission can be the source of re-entrant code
+          // in functional libraries like this. We only really
+          // need to do this if it's the first value, or if the
+          // key we're tracking in previous needs to change.
+          first = false;
+          previousKey = currentKey;
+
+          // Emit the value!
           subscriber.next(value);
         }
-        prev = key;
-        first = false;
```

Comment:
These can't happen _after_ we next, it has to happen before. Otherwise the added test will fail.